### PR TITLE
feat(UIPatternBlock): implement UI pattern block

### PR DESCRIPTION
**Pull request by fulopdaniel**

### CODE REVIEW

File: packages/ui-pattern-block/src/UIPatternBlock.tsx [hunk 1]
```diff
@@ -1,25 +1,238 @@
...
+        }
+    }, [dependencies]);
+
+    const parsedExternalDependencies = useMemo(() => {
+        try {
+            return JSON.parse(externalDependencies);
+        } catch {
+            return [];
+        }
+    }, [dependencies]);
```
Comment (ragi96) [inline packages/ui-pattern-block/src/UIPatternBlock.tsx:141]: They are both almost the same, can we have them like that:

```
const useParsedDependencies = (dependenciesString: string) => {
    return useMemo(() => {
        try {
            return JSON.parse(dependenciesString);
        } catch {
            return {};
        }
    }, [dependenciesString]);
};
```

And then:

```
const parsedNpmDependencies = useParsedDependencies(npmDependencies);
const parsedExternalDependencies = useParsedDependencies(externalDependencies);
```

File: packages/ui-pattern-block/src/UIPatternBlock.tsx [hunk 2]
```diff
@@ -1,25 +1,238 @@
...
+    );
+
+    const onResetToDefault = () => {
+        setReset(!reset);
+    };
+
+    const onCodeChange = (filename: string, code: string) => {
+        if (!isEditing) {
+            return;
+        }
```
Comment (ragi96) [inline packages/ui-pattern-block/src/UIPatternBlock.tsx:89]: Why is this needed?

File: packages/ui-pattern-block/src/helpers/style.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
```
Comment (ragi96) [inline packages/ui-pattern-block/src/helpers/style.ts:1]: Add tests for these functions

File: packages/ui-pattern-block/src/helpers/style.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import { Radius, getBackgroundColorStyles, radiusStyleMap } from '@frontify/guideline-blocks-settings';
+import { Height, Padding, heightValues, paddingValues } from '../types';
+import { Color } from '@frontify/fondue';
```
Comment (ragi96) [inline packages/ui-pattern-block/src/helpers/style.ts:5]:
```suggestion
import type { Color } from '@frontify/fondue';
```

File: packages/ui-pattern-block/src/hooks/useDebounce.tsx [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
```
Comment (ragi96) [inline packages/ui-pattern-block/src/hooks/useDebounce.tsx:1]: Can't we use the `useDebounce` of `fondue`?

File: packages/ui-pattern-block/src/helpers/files.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { SandpackFiles } from '@codesandbox/sandpack-react/types';
```
Comment (ragi96) [inline packages/ui-pattern-block/src/helpers/files.ts:3]:
```suggestion
import type { SandpackFiles } from '@codesandbox/sandpack-react/types';
```

File: packages/ui-pattern-block/src/components/Accordion.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+export const Accordion = ({ label, children, isOpen, setIsOpen }: PropsWithChildren<Props>): ReactElement => {
+    return (
+        <div data-test-id="dependency-accordion" className="tw-border-b tw-border-b-line last:tw-border-b-0">
+            <button
+                className={joinClassNames([
+                    'tw-text-s tw-gap-2 tw-w-full tw-text-text-weak tw-box-content tw-bg-white tw-h-10 tw-px-4 tw-flex tw-items-center',
+                    isOpen && 'tw-border-b tw-border-b-line',
+                ])}
+                onClick={() => setIsOpen(!isOpen)}
+            >
```
Comment (ragi96) [inline packages/ui-pattern-block/src/components/Accordion.tsx:22]:
```suggestion
            <button
                className={joinClassNames([
                    'tw-text-s tw-gap-2 tw-w-full tw-text-text-weak tw-box-content tw-bg-white tw-h-10 tw-px-4 tw-flex tw-items-center',
                    isOpen && 'tw-border-b tw-border-b-line',
                ])}
                onClick={() => setIsOpen(!isOpen)}
                ariaExpanded={isOpen}
            >
```

File: packages/ui-pattern-block/src/components/Captions.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            <RichTextEditor
+                id={`${appBridge.getBlockId().toString()}_description`}
+                isEditing={isEditing}
+                value={description ?? convertToRteValue(TextStyles.imageCaption, '')}
+                plugins={getCaptionPlugins(appBridge)}
+                onTextChange={onDescriptionChange}
+                showSerializedText={hasRichTextValue(description)}
+                placeholder="Add a description here"
+            />
+        </div>
```
Comment (ragi96) [inline packages/ui-pattern-block/src/components/Captions.tsx:52]: NIT but we could group common props:

```suggestion
     const commonProps = {
        isEditing,
        plugins: getCaptionPlugins(appBridge),
    };
    return (
        <div className={joinClassNames(['tw-flex tw-flex-col tw-gap-1', shouldRender && 'tw-mb-3'])}>
            <RichTextEditor
                id={`${appBridge.getBlockId().toString()}_title`}
                value={title ?? convertToRteValue(TextStyles.imageTitle, '')}
                onTextChange={onTitleChange}
                showSerializedText={hasRichTextValue(title)}
                placeholder="Pattern name"
            />
            <RichTextEditor
                id={`${appBridge.getBlockId().toString()}_description`}
                value={description ?? convertToRteValue(TextStyles.imageCaption, '')}
                onTextChange={onDescriptionChange}
                showSerializedText={hasRichTextValue(description)}
                placeholder="Add a description here"
            />
        </div>
```

File: packages/ui-pattern-block/src/components/ResponsivePreview.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
```
Comment (ragi96) [inline packages/ui-pattern-block/src/components/ResponsivePreview.tsx:1]: Can scroll the page when its open

File: packages/ui-pattern-block/src/components/ToolbarButton.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+            hoverDelay={75}
+            enterDelay={75}
+            withArrow
+            triggerElement={
+                <Button
+                    data-test-id="toolbar-icon-btn"
+                    onClick={onClick}
+                    icon={icon}
+                    size={ButtonSize.Small}
+                    emphasis={ButtonEmphasis.Weak}
```
Comment (ragi96) [inline packages/ui-pattern-block/src/components/ToolbarButton.tsx:26]: Missing `aria-label` and `aria-labelledby attribute cannot be used on a div with no valid role attribute.`

File: packages/ui-pattern-block/src/UIPatternBlock.tsx [hunk 11]
```diff
@@ -1,25 +1,238 @@
...
+                                    ...getBackgroundColorStyles(backgroundColor),
+                                    backgroundImage: 'none',
+                                }),
+                            }}
+                            showRefreshButton={false}
+                            showOpenInCodeSandbox={false}
+                            className="tw-rounded-none tw-bg-box-nega tw-shadow-none tw-ml-0"
+                        />
+                        {(isEditing || showCode) && (
+                            <CodeEditor
```
Comment (ragi96) [inline packages/ui-pattern-block/src/UIPatternBlock.tsx:203]: Can't access the CodeEditor by keyboard

File: packages/ui-pattern-block/src/components/ResponsivePreview.tsx (hunk 9 above)
Comment (ragi96) [inline packages/ui-pattern-block/src/components/ResponsivePreview.tsx:1]: Can't change the preview device by keyboard only

File: packages/ui-pattern-block/manifest.json [hunk 12]
```diff
@@ -1,3 +1,3 @@
 {
-    "appId": "clmyyhwdf00010nuqq138mw3s"
+    "appId": "clncxpkhl00014muq6fm3u7i0"
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/manifest.json]: I will remove the old versions, but I want to keep them for a bit for comparison purposes if needed. Btw old blocks were separated - one for react and one for vanilla. Thought it made sense to create a new for this new one.

File: packages/ui-pattern-block/src/UIPatternBlock.spec.ct.tsx [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            editorState: false,
+            blockId: 6,
+            blockSettings: { ...DEFAULT_BLOCK_SETTINGS, showCode: false },
+        });
+        mount(<UIPatternBlockWithStubs />);
+        cy.get(CodeEditorSelector).should('not.exist');
+        cy.get(ToolbarSelector).should('not.exist');
+    });
+
+    it('collapses the code editor if it is set in the settings', () => {
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.spec.ct.tsx:163]: Added 👍

File: packages/ui-pattern-block/src/UIPatternBlock.spec.ct.tsx [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+        const [UIPatternBlockWithStubs] = withAppBridgeBlockStubs(UIPatternBlock, {
+            editorState: false,
+            blockId: 12,
+            blockSettings: { ...DEFAULT_BLOCK_SETTINGS, showSandboxLink: false, showResetButton: false },
+        });
+        mount(<UIPatternBlockWithStubs />);
+        cy.get(ToolbarIconButtonSelector).should('have.length', 2);
+    });
+
+    it('opens the responsive preview on button click and changes preview width', () => {
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.spec.ct.tsx:259]: Added

File: packages/ui-pattern-block/.eslintrc.js [hunk 15]
```diff
@@ -21,6 +21,7 @@ module.exports = {
                         },
                     },
                 ],
+                'react-hooks/exhaustive-deps': 'off',
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/.eslintrc.js]: Done

File: packages/ui-pattern-block/manifest.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/manifest.json]: Added

File: packages/ui-pattern-block/package.json [hunk 16]
```diff
@@ -17,6 +17,7 @@
         "@babel/core": "^7.22.17",
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/package.json:17]: Updated

File: packages/ui-pattern-block/src/UIPatternBlock.tsx [hunk 17]
```diff
@@ -1,25 +1,238 @@
...
+import { useBlockSettings, useEditorState } from '@frontify/app-bridge';
+import { BlockProps, getBackgroundColorStyles, getBorderStyles } from '@frontify/guideline-blocks-settings';
+import { SandpackLayout, SandpackPreview, SandpackProvider } from '@codesandbox/sandpack-react';
 
 import 'tailwindcss/tailwind.css';
 import '@frontify/guideline-blocks-settings/styles';
 
-import { type Settings, sandpackThemeValues } from './types';
+import { Alignment, Height, type Settings, sandpackThemeValues } from './types';
+import { CodeEditor } from './components/CodeEditor';
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx]: Good catch, fixed

File: packages/ui-pattern-block/src/UIPatternBlock.tsx [hunk 18]
```diff
@@ -1,25 +1,238 @@
...
+    getCssToInject,
+    getDefaultFilesOfTemplate,
+    getHeightStyle,
+    getPaddingStyle,
+    getRadiusStyle,
+    getScriptToInject,
+    initialActiveFile,
+} from './helpers';
+import { useDebounce } from './hooks';
+import { Captions, ExternalDependencies, NPMDependencies } from './components';
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx]: Yep, thanks!

File: packages/ui-pattern-block/src/UIPatternBlock.tsx (hunk 1 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx]: Created a helper function for these.

File: packages/ui-pattern-block/src/UIPatternBlock.tsx (hunk 2 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx:89]: so that the changes that viewers make are not saved to the database.

File: packages/ui-pattern-block/src/hooks/useDebounce.tsx (hunk 5 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/hooks/useDebounce.tsx:1]: I tried it but that was not working properly when I used it in a function in a component.

File: packages/ui-pattern-block/src/components/Accordion.tsx (hunk 7 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/Accordion.tsx]: added 👍

File: packages/ui-pattern-block/src/components/Captions.tsx (hunk 8 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/Captions.tsx]: Done 👍

File: packages/ui-pattern-block/src/components/ToolbarButton.tsx (hunk 10 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/ToolbarButton.tsx]: Added aria-label. AFAIK aria-labelledby is not needed here.

File: packages/ui-pattern-block/src/UIPatternBlock.tsx (hunk 11 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx:201]: Fixed.

File: packages/ui-pattern-block/src/components/ResponsivePreview.tsx (hunk 9 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/ResponsivePreview.tsx:1]: Fixed

File: packages/ui-pattern-block/manifest.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (ragi96) [reply packages/ui-pattern-block/manifest.json]: Well the one that was set I created after the settings task ;)

File: packages/ui-pattern-block/src/UIPatternBlock.tsx [hunk 19]
```diff
@@ -1,25 +1,239 @@
...
+        const alignmentCss = alignment === Alignment.Center ? centeringCss : '';
+        const backgroundCss = hasBackground ? getBackgroundCss(backgroundColor) : '';
+        const hasAutoHeight = !isCustomHeight && heightChoice === Height.Auto;
+        return getScriptToInject(getCssToInject(hasAutoHeight, alignmentCss, backgroundCss));
+    }, [hasBackground, backgroundColor, alignment, isCustomHeight, heightChoice]);
+
+    const templateFiles = useMemo(
+        () => ({ ...getDefaultFilesOfTemplate(sandpackTemplate), ...files?.[sandpackTemplate] }),
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        [sandpackTemplate, cssToInject, dependencies, reset],
```
Comment (ragi96) [inline packages/ui-pattern-block/src/UIPatternBlock.tsx:79]:
```suggestion
        [sandpackTemplate, files],
```

What I am missing that its not like that?

File: packages/ui-pattern-block/src/components/JsonErrorMessage.tsx [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+import { ReactElement } from 'react';
+
+export const JsonErrorMessage = (): ReactElement => {
+    return (
+        <div
+            data-test-id="dependency-invalid-json"
+            className="tw-text-box-negative-inverse tw-w-full tw-h-11 tw-flex tw-items-center tw-px-3 tw-gap-3 tw-bg-box-negative tw-text-s tw-font-medium tw-font-sans"
+        >
+            <IconExclamationMarkTriangle20 />
+            <span>The entered JSON is invalid</span>
```
Comment (ragi96) [inline packages/ui-pattern-block/src/components/JsonErrorMessage.tsx:13]: I think we need to announce that error message to screen readers for accessibility, probably with a `aria-live="assertive"`, would be nice if you could ask @LidijaStoilova about that

File: packages/ui-pattern-block/src/components/CodeEditor.tsx [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+}: Props): ReactElement => {
+    const { code } = useActiveCode();
+    const { sandpack } = useSandpack();
+    const { activeFile } = sandpack;
+    const [isEditorCollapsed, setIsEditorCollapsed] = useState(shouldCollapseCodeByDefault);
+
+    useEffect(() => {
+        onCodeChange(activeFile, code);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [code]);
```
Comment (ragi96) [inline packages/ui-pattern-block/src/components/CodeEditor.tsx:37]: This one seems odd, as there is a bug and I can't adjust it. As the adjustment got stored and then another call will be made with the old code and overwrite the change.

File: packages/ui-pattern-block/src/UIPatternBlock.tsx (hunk 11 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx:216]: With the escape key. Unfortunately the content manager also listens to escape and unfocuses the block.

This PR will fix that: https://github.com/Frontify/web-app/pull/1542

File: packages/ui-pattern-block/manifest.json [hunk 22]
```diff
@@ -1,3 +1,7 @@
 {
-    "appId": "clmyyhwdf00010nuqq138mw3s"
+    "appId": "clncxpkhl00014muq6fm3u7i0",
+    "searchFields": [
+        { "type": "fondueRte", "settingId": "title" },
+        { "type": "fondueRte", "settingId": "description" }
+    ]
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/manifest.json]: Thanks ,added

File: packages/ui-pattern-block/src/components/ResponsivePreview.tsx [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+                    lastFocusableElement.focus();
+                    e.preventDefault();
+                }
+            }
+        };
+        window.addEventListener('keydown', keyHandler);
+        return () => window.removeEventListener('keydown', keyHandler);
+    }, [onPreviewClose]);
+
+    return createPortal(
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/ResponsivePreview.tsx:66]: I checked but unless I am missing something, creating a layout that is required here is not possible with the fondue modal.

File: packages/ui-pattern-block/src/components/ResponsivePreview.tsx [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+                                        ])}
+                                    >
+                                        {device.label}
+                                    </button>
+                                    {i !== devices.length - 1 && <div className="tw-bg-line tw-w-[1px] tw-h-full" />}
+                                </Fragment>
+                            ))}
+                        </div>
+                        <div className="tw-absolute tw-right-0">
+                            <Button
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/ResponsivePreview.tsx:97]: thanks, added!

File: packages/ui-pattern-block/src/UIPatternBlock.tsx (hunk 19 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx]: so every time this variable changes the sandpack provider is remounted. That's how it works internally. If the `files` variable was added to the dependency array, it would remount on every keystroke. 

The other dependencies are needed, because when e.g. dependencies change, then the sandbox is remounted, and we need to get the latest saved files, so we need to rerun the usememo.

I changes the dependency array to be the same array I use for the key attribute, so they always match.

File: packages/ui-pattern-block/src/components/JsonErrorMessage.tsx (hunk 20 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/JsonErrorMessage.tsx]: Thanks ,added

File: packages/ui-pattern-block/src/components/CodeEditor.tsx (hunk 21 above)
Comment (fulopdaniel) [reply packages/ui-pattern-block/src/components/CodeEditor.tsx:43]: Fixed!! Thanks, great catch

**Review by ragi96 — COMMENTED** [review]
Looks good, but there are some things we need to consider before we merge it:

1. deactivate the layout on the preview, as it will look odd the most times:
Screenshot 2023-10-18 at 14 21 22
2. Sass / Less support?
3. Typescript also as a template?
4. Just showing the "responsive view" button on desktop?
5. Adding the SandpackConsole? (optionally)

File: packages/ui-pattern-block/src/UIPatternBlock.tsx (hunk 11 above)
Comment (ragi96) [reply packages/ui-pattern-block/src/UIPatternBlock.tsx:216]: Works as expected with the open PR

**Review by ragi96 — APPROVED** [review]
